providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+        ref_lane = ((rnd >> 32)) % ctx->lanes;
+        /* Can not reference other lanes yet */
+        if ((pass == 0) && (slice == 0))
+            ref_lane = lane;
+
+        /* Computing the number of possible reference block within the lane. */
+        index = j;
+        ref_index = index_alpha(ctx, pass, lane, slice, j, rnd & 0xFFFFFFFF,
+                                ref_lane == lane);
+
+        /* Creating a new block */
+        ref_block = ctx->memory + ctx->lane_length * ref_lane + ref_index;
+        curr_block = ctx->memory + curr_offset;
+        if (ARGON2_VERSION_10 == ctx->version) {
+            /* Version 1.2.1 and earlier: overwrite, not XOR */
+            fill_block(ctx->memory + prev_offset, ref_block, curr_block, 0);
+            continue;
+        }
+
+        if(pass == 0)
+            fill_block(ctx->memory + prev_offset, ref_block, curr_block, 0);
+        else
+            fill_block(ctx->memory + prev_offset, ref_block, curr_block, 1);
+    }
+}
+
+
+
+#if !defined(ARGON2_NO_THREADS) && defined(OPENSSL_THREADS)
+
+static uint32_t fill_segment_thr(void *thread_data)
+{
+    ARGON2_THREAD_DATA *my_data;
+
+    my_data = (ARGON2_THREAD_DATA *) thread_data;
+    fill_segment(my_data->ctx, my_data->pos.pass, my_data->pos.lane, my_data->pos.slice, my_data->pos.index);
+    return 0;
+}
+
+static int fill_mem_blocks_mt(ARGON2_CTX *ctx) {
+    uint32_t r, s, l, ll;
+    void **t;
+    ARGON2_THREAD_DATA *t_data;
+
+    t = OPENSSL_zalloc(sizeof(void*)*ctx->lanes);
+    t_data = OPENSSL_zalloc(ctx->lanes*sizeof(ARGON2_THREAD_DATA));
+
+    if (t == NULL || t_data == NULL)
+        goto fail;
+
+    for (r = 0; r < ctx->passes; ++r) {
+        for (s = 0; s < ARGON2_SYNC_POINTS; ++s) {
+            for (l = 0; l < ctx->lanes; ++l) {
+                ARGON2_POS p;
+                if (l >= ctx->threads)
+                    if (crypto_thread_join(ctx->openssl_ctx, t[l - ctx->threads],
+                                           NULL) == 0)
+                        goto fail;
+
+                p.pass = r;
+                p.lane = l;
+                p.slice = (uint8_t)s;
+                p.index = 0;
+
+                t_data[l].ctx = ctx;
+                memcpy(&(t_data[l].pos), &p, sizeof(ARGON2_POS));
+                t[l] = crypto_thread_start(ctx->openssl_ctx, &fill_segment_thr,
+                                           (void*) &t_data[l]);
+                if (t[l] == NULL) {
+                    for (ll = 0; ll < l; ++ll)
+                        crypto_thread_join(ctx->openssl_ctx, t[ll], NULL);
+                    t[ll] = NULL;
+                    goto fail;
+                }
+            }
+            for (l = ctx->lanes - ctx->threads; l < ctx->lanes; ++l) {
+                if (crypto_thread_join(ctx->openssl_ctx, t[l], NULL) == 0)
+                    goto fail;
+                t[l] = NULL;
+            }
+        }
+    }
+
+    crypto_thread_clean(ctx->openssl_ctx, NULL);
```

Comment:
What would happen here if something else called crypto_thread_clean(ctx->openssl_ctx, NULL) during this function??

Reply:
crypto_thread_clean locks a mutex to avoid issues of that sort.